### ISSUE TITLE
A more basic usage example?

### ISSUE THREAD

**Issue by talgalili** [shabbychef/PDQutils#2]
Dear Steven,
Thank you for your interesting R package.

I am looking for a method to fit an approximate PDF (i.e.: density estimation) using the first k (empirical) moments (see my SO question here: http://stats.stackexchange.com/questions/175323/how-to-fit-an-approximate-pdf-i-e-density-estimation-using-the-first-k-empi).

I was wondering if you would be willing to provide an example for such a use case.

Thanks!
Tal

**Comment by shabbychef** [comment]
Convert the raw empirical moments to cumulants, then use the Edgeworth expansion, which requires cumulants. So something like:

``` r
x <- empirical_samples_of_my_rv(1e4)
raw_moments <- vapply(c(1:4),FUN.VALUE=0,function(n) { mean(abs(x)^n) })
raw_cumulants <- PDQutils::moment2cumulant(raw_moments)
xseq <- seq(min(x),max(x),length.out=1e3)
adens <- PDQutils::dapx_edgeworth(xseq,raw_cumulants)
plot(xseq,adens)
```

Note that typically the Edgeworth expansions do a better job (converge as you add terms), but I have not coded the non-normal case. This is simpler to do for the Gram Charlier code, and I have added the ability to model distributions which are more 'like' a gamma or a beta and so on. For these, you do not even need to compute the cumulants, since they take the raw moments:

``` r
adens <-PDQutils::dapx_gca(xseq,raw_moments,basis='gamma')
plot(xseq,adens)
```

hth, I will add this to the vignette/README.

FWIW, the Edgeworth functions should probably optionally take the raw moments and do the conversion for you. I will make that issue #3.

**Comment by talgalili** [comment]
Dear @shabbychef , thank you for the quick/detailed answer.

I ran a quick test, and the estimated results does not look "so good". Is this to be expected?

``` r
set.seed(2015-10-05)
empirical_samples_of_my_rv <- function(n) rnorm(n)

x <- empirical_samples_of_my_rv(1e5)
n_moments <- 5
raw_moments <- vapply(c(1:n_moments),FUN.VALUE=0,function(n) { mean(abs(x)^n) })
raw_cumulants <- PDQutils::moment2cumulant(raw_moments)
xseq <- seq(min(x),max(x),length.out=1e3)
adens <- PDQutils::dapx_edgeworth(xseq,raw_cumulants)
plot(xseq, adens, type = "l")
lines(density(x), col = 2)
```

Result (the original standard normal distribution in red, the black line is the estimated density):

![image](https://cloud.githubusercontent.com/assets/976006/10278945/19562194-6b6a-11e5-8169-000459ec56de.png)

**Comment by talgalili** [comment]
After further checks, I can see this has to do with the negative values. Is that solvable?

**Comment by shabbychef** [comment]
I am so embarrassed. The computation of moments was wrong in my example. Try:

``` r
set.seed(2015-10-05)
empirical_samples_of_my_rv <- function(n) rnorm(n)

x <- empirical_samples_of_my_rv(1e5)
n_moments <- 5
# not this:
#raw_moments <- vapply(c(1:n_moments),FUN.VALUE=0,function(n) { mean(abs(x)^n) })
# this:
raw_moments <- vapply(c(1:n_moments),FUN.VALUE=0,function(n) { mean((x)^n) })
raw_cumulants <- PDQutils::moment2cumulant(raw_moments)
xseq <- seq(min(x),max(x),length.out=1e3)
adens <- PDQutils::dapx_edgeworth(xseq,raw_cumulants)
plot(xseq, adens, type = "l")
lines(density(x), col = 2)
```

**Comment by talgalili** [comment]
github is for debugging :)

IF I may ask, in layman's terms, what are the reasonable limitations of this method?
The distribution needs to be unimodal? how many moments can we reasonably use? How well does it deal with things like log-normal distributions?

Again, great work, and thank you for the responsiveness!
p.s.: if you'd like to set up a guest post on r-statistics.com / r-bloggers.com, I would gladly oblige.
Tal

**Comment by shabbychef** [comment]
I am by no means an 'expert' in this area; I only wrote the package because I needed the functionality. That said, I would refer to the paper by [Blinnikov & Moessner](http://arxiv.org/abs/astro-ph/9711239), who give some vague convergence results (in the number of terms used) for the different methods. I suspect that unimodality is _not_ technically required (rather it seems to be convergence of higher order cumulants).

In my experience, using an appropriate base distribution helps quite a bit: use the (default) normal for unbounded unimodal distributions, gamma for one-sided distributions like chi-squares, and beta for bounded distributions. The counterexamples in Blinnikov & Moessner focus on using a normal base distribution to model a chi-square (and failing at it). I think using the proper base distribution is also a bigger win than using a monotonic rearrangement (the paper by Chernozhukov _et. al._ seems to be a solution in search of a problem). I have not yet incorporated different base distributions to the Edgeworth expansions, however, only to the Gram Charlier expansions; the latter are easier for a former numerical analyst like myself to understand. The paper by [Berberan Santos](http://web.ist.utl.pt/ist12219/data/115.pdf) essentially gives the algorithm.

hth; I may take you up on your offer at some point, but am rather busy right now.

**Comment by talgalili** [comment]
Hi Steven,
Thank you for all the info.
Packing it all up for a nice guest post could be great. Whenever you are up to it, give me a ping: tal.galili@gmail.com

Cheers,
Tal
